### PR TITLE
Use QgsTextRenderer to render attribute table text

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows for full range of text formatting in layout attribute tables, including buffers & shadows, word spacing, etc

Sponsored by City of Canning

### CODE REVIEW

File: src/core/layout/qgslayouttable.cpp [hunk 1]
```diff
@@ -612,31 +697,62 @@ void QgsLayoutTable::setShowEmptyRows( const bool showEmpty )
 
 void QgsLayoutTable::setHeaderFont( const QFont &font )
 {
-  if ( font == mHeaderFont )
```
Comment (nirvn) [inline src/core/layout/qgslayouttable.cpp:615]: @nyalldawson , any reason why you're not doing a font == mHeaderFont.toQFont() check here?

File: src/core/layout/qgslayouttable.cpp [hunk 2]
```diff
@@ -665,31 +781,62 @@ void QgsLayoutTable::setHeaderMode( const QgsLayoutTable::HeaderMode mode )
 
 void QgsLayoutTable::setContentFont( const QFont &font )
 {
-  if ( font == mContentFont )
```
Comment (nirvn) [inline src/core/layout/qgslayouttable.cpp:668]: See question above.

File: src/core/layout/qgslayouttable.h [hunk 3]
```diff
@@ -628,40 +662,42 @@ class CORE_EXPORT QgsLayoutTable: public QgsLayoutMultiFrame
...
      * \param drawHeaderLines set to TRUE to include for the table header
      * \see drawVerticalGridLines()
      */
-    void drawHorizontalGridLines( QPainter *painter, int firstRow, int lastRow, bool drawHeaderLines ) const;
+    void drawHorizontalGridLines( QgsLayoutItemRenderContext &context, int firstRow, int lastRow, bool drawHeaderLines ) const;
 
     /**
      * Draws the vertical grid lines for the table.
-     * \param painter destination painter for grid lines
+     * \param destination render context
```
Comment (nirvn) [inline src/core/layout/qgslayouttable.h:700]:
```suggestion
     * \param context destination render context
```

File: src/gui/layout/qgslayoutattributetablewidget.cpp [hunk 4]
```diff
@@ -72,8 +70,8 @@ QgsLayoutAttributeTableWidget::QgsLayoutAttributeTableWidget( QgsLayoutFrame *fr
   connect( mUseConditionalStylingCheckBox, &QCheckBox::stateChanged, this, &QgsLayoutAttributeTableWidget::useConditionalStylingChanged );
   setPanelTitle( tr( "Table Properties" ) );
 
-  mContentFontToolButton->setMode( QgsFontButton::ModeQFont );
-  mHeaderFontToolButton->setMode( QgsFontButton::ModeQFont );
+  mContentFontToolButton->setMode( QgsFontButton::ModeTextRenderer );
```
Comment (nirvn) [inline src/gui/layout/qgslayoutattributetablewidget.cpp:73]: :tada:

File: src/gui/qgsfontbutton.cpp [hunk 5]
```diff
@@ -869,6 +869,7 @@ void QgsFontButton::updatePreview( const QColor &color, QgsTextFormat *format, Q
       context.setUseAdvancedEffects( true );
       context.setFlag( QgsRenderContext::Antialiasing, true );
       context.setPainter( &p );
+      context.setFlag( QgsRenderContext::Antialiasing, true );
```
Comment (nirvn) [inline src/gui/qgsfontbutton.cpp:872]: You set the antialiasing flag twice over three lines, should one setFlag() go away?

Comment (nyalldawson) [reply src/gui/qgsfontbutton.cpp:872]: Ah, bad rebase sorry, fixed

File: src/core/layout/qgslayouttable.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/layout/qgslayouttable.cpp:615]: I'd also have to check the logic regarding size units too.. seems like overkill for a deprecated method which isn't used anywhere anymore!

**Comment by nyalldawson** [comment]
Thanks @nirvn!

File: src/core/layout/qgslayouttable.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/layout/qgslayouttable.cpp:615]: @nyalldawson , fair enough.

**Pull request merged**